### PR TITLE
Gate Start/Rebase on dependency-aware base freshness

### PR DESCRIPTION
**Pull request by subsetpark**
## Problem

Tracing a real run (`ts-lowerability-effect-boundary-cleanup`, where patch 5 → {1,2,3} re-implemented patch 3 byte-for-byte) surfaced two related defects in how patch branches are cut and rebased:

1. **Fan-in blind spot.** When a patch `P` depends on several patches and exactly one is still open, `P` is cut from that sole open dep `B` (`Graph.initial_base`). `B` carries its own lineage but **not** a *sibling* dependency of `P` that squash-merged to `main` independently — that squash commit lives on `main`, never an ancestor of `B`. So `P` starts from a base missing one of its dependencies, and the agent re-implements it. The old start gate (`Start_eligibility`) only checked `B`'s *own* deps, so it was blind to `P`'s siblings.

2. **No topological ordering of the rebase cascade.** `runnable_messages` explicitly did not gate `Rebase` on base freshness (*"Rebase is what closes the freshness gap"*). For a chain `B ← C ← E`, `C` could rebase onto `B` while `B` was itself stale or mid-rebase, landing on a stale base.

## Approach

A **sibling-containment gate** plus **topological rebase ordering**, keeping the pure-decision (`lib_core`) / effectful-handler (`lib`) split:

- **`Base_containment.contains_merged_siblings`** (new pure module): does the patch's resolved base contain the squash commit of every *merged* dependency? Fed by an injected `ancestor_oracle` (same pattern as `Rebase_decision.plan`). Fail-closed on unknown SHA; trivially true for base = `main`; **only the patch's own deps are consulted**, so unrelated `main` movement never trips it.
- **Merge-commit SHA plumbing**: fetch GitHub `mergeCommit.oid` → `Pr_state` → `Poller` → recorded on `Patch_agent` when merged. Persisted (merged agents aren't re-polled) with a self-heal that keeps polling a merged agent until its SHA is known.
- **`poller_fiber`** recomputes a `base_contains_merged_siblings` cache each tick via `git merge-base --is-ancestor`, against `initial_base` (the actual cut base, not the `None` pre-start `base_branch`).
- **`Start_eligibility.decide`** gains the containment input + a `Base_missing_merged_sibling` defer reason (checked last; `main`/merged already imply containment).
- **`reconciler.detect_sibling_stale_bases`** enqueues a rebase of the **sole open dep `B`** — the demand the existing three detectors miss, since `B` is a *sibling* of the merged dep, not a dependent — deduped with them.
- **`runnable_messages` now gates BOTH `Start` and `Rebase`** on the unified predicate. This reverses the old "Rebase is not freshness-sensitive" decision and makes a fan-in/chain cascade **block on its dependency layers and converge bottom-up** (`main` is always fresh), one rebase per branch, never onto a stale base.

## Tests

- **`test_base_containment.ml`** (new): every fan-in permutation, fail-closed-on-missing-SHA, and the "unrelated merges never flip containment" invariant — pure, via a mocked oracle.
- **`test_start_eligibility_properties.ml`**: new-arm reachability + precedence (`Base_missing_merged_sibling`), strengthened SEP-4/SEP-5.
- **`test_reconciler_properties.ml`**: `detect_sibling_stale_bases` matrix — the three real-bug permutations targeting the correct base, silent/idempotent/off-edge cases, and the reconcile-level demand.

All pre-existing suites (incl. inline `let%test`s and `stale_base_invariants`) pass under the new gate. `dune build`, `dune test`, and `dune build @fmt` are green.

## Note for reviewers

The pure decision logic and detector are exhaustively property-tested, and the containment computation is unit-tested via the extracted oracle-injected function. A real-git end-to-end test exercising `is_ancestor` + `merge_commit_sha` through the live `poller_fiber` functor was **not** added (heavyweight to wire); the git-subprocess layer remains covered by the existing `Worktree.is_ancestor` tests.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Gates Start and Rebase on dependency‑aware base freshness so fan‑in patches don’t cut from bases missing merged sibling deps and cascades run bottom‑up. Adds a fail‑closed, pure containment check with lazy git I/O, plumbs/persists `mergeCommit.oid` (incl. prune path), and keeps polling merged PRs until the SHA is known; older snapshots default fail‑closed.

- **Bug Fixes**
  - Compute a `base_contains_merged_siblings` cache each poll via `git merge-base --is-ancestor`; fail‑closed on missing SHA and transient git errors; lazy main root; keep polling merged PRs without a SHA.
  - Gate both Start and Rebase on this predicate so cascades block on deps and converge bottom‑up (no rebases onto stale or mid‑rebase bases).
  - Add a sibling‑stale detector that enqueues a Rebase for the sole open dependency; only if that base has a PR and no Rebase queued; deduplicated across detectors and tightened for invariant edge cases.
  - Plumb and persist GitHub `mergeCommit.oid`; record it on the prune path to avoid waiting for a re‑poll.
  - Persistence: default `base_contains_merged_siblings` to `false` when absent in older snapshots to avoid fail‑open on load.

<sup>Written for commit df3799b0c1aace968b890e61e677c9eae25930cf. Summary will update on new commits. <a href="https://cubic.dev/pr/flowglad/onton/pull/332?utm_source=github">Review in cubic</a></sup>

## Summary by CodeRabbit

* **New Features**
  * Added automatic rebase detection for fan-in patches when merged sibling dependencies are missing from the base branch.
  * Implemented merge commit SHA tracking to improve ancestry verification and dependency resolution.
  * Enhanced eligibility gating to prevent operations when bases lack required merged sibling commits.

* **Improvements**
  * Strengthened handling of complex multi-dependency patch scenarios through improved base containment checking.

[Review Change Stack](https://app.coderabbit.ai/change-stack/flowglad/onton/pull/332?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: lib/poller_fiber.ml [hunk 1]
```diff
@@ -431,6 +443,41 @@ struct
...
+              match Orchestrator.find_agent orch pid with
+              | Some a -> a.Patch_agent.merge_commit_sha
+              | None -> None
+            in
+            let main_root = W.resolve_main_root () in
+            let ancestor_oracle ancestor ~descendant =
+              W.is_ancestor ~path:main_root ~ancestor ~descendant
+            in
+            let orch =
+              List.fold (Orchestrator.all_agents orch) ~init:orch
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/poller_fiber.ml:489]: Fixed in \`594a048\`. Confirmed it really can raise: \`W.is_ancestor\` → \`run_git_exit_code\` → \`retry_transient_spawn\`, which re-raises after exhausting retries. The oracle now does \`try W.is_ancestor … with Eio.Cancel.Cancelled _ as e -> raise e | _ -> false\`, so a transient git spawn failure fails closed instead of aborting the reconcile tick — matching the fail-closed contract in the \`.mli\`.

File: lib/poller_fiber.ml [hunk 2]
```diff
@@ -431,6 +443,41 @@ struct
...
+               calls are limited to fan-in patches mid-merge. Fail-closed to
+               [false] when a dep's merge SHA is not yet recorded — the
+               dependent then defers until [detect_sibling_stale_bases] rebases
+               the base (which also fetches the merge commit into the repo). *)
+            let graph = Orchestrator.graph orch in
+            let has_merged pid =
+              match Orchestrator.find_agent orch pid with
+              | Some a -> a.Patch_agent.merged
+              | None -> false
+            in
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/poller_fiber.ml:460]: Fixed in \`594a048\`. \`main_root\` is now \`lazy\`, and \`Base_containment.contains_merged_siblings\` short-circuits (returns before any oracle call) for a base on \`main\` or a patch with no merged deps. So a project with no fan-in-mid-merge performs zero git I/O here.

File: lib/persistence.ml [hunk 3]
```diff
@@ -308,6 +311,11 @@ let patch_agent_of_yojson ~gameplan json =
        ~ci_failure_count:(int_member "ci_failure_count" json)
        ~session_fallback ~human_messages ~inflight_human_messages ~ci_checks
        ~merge_ready:(bool_member "merge_ready" json)
+       ~merge_commit_sha:(string_member_opt "merge_commit_sha" json)
+       ~base_contains_merged_siblings:
+         (Option.value
+            (bool_member_opt "base_contains_merged_siblings" json)
+            ~default:true)
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/persistence.ml]: Applied in \`187c799\`. The missing-key default is now \`false\` (fail-closed), so loading a pre-field snapshot cannot fail-open the gate for a fan-in patch. Round-trip identity is preserved because snapshots written by this version always include the key (the default is only hit for old snapshots); the first poll recomputes the live value regardless.

File: lib_core/base_containment.ml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+open Base
+
+let contains_merged_siblings ~graph ~patch_id ~has_merged ~merge_sha ~branch_of
+    ~main ~ancestor_oracle =
+  match Graph.open_pr_deps graph patch_id ~has_merged with
+  | _ :: _ :: _ ->
+      (* Not at the last-but-one-merge edge; the patch is not startable and
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib_core/base_containment.ml:7]: Verified: \`Graph.initial_base\` (graph.ml:65) is total for ≤1 open dep — \`[] -> main\` (including a root patch with no deps) and \`[d] -> branch_of d\`; it only raises on \`>1\`, which the \`_ :: _ :: _\` arm above excludes, so the call never raises. Added a comment citing this in \`187c799\`.

File: lib/orchestrator.ml [hunk 5]
```diff
@@ -324,15 +327,26 @@ let runnable_messages t =
...
-             and [Respond] operates on a worktree that already exists. *)
+          (* Gate both [Start] and [Rebase] on base freshness. A [Rebase] of a
+             dependent must not run while its base is itself stale or mid-rebase
+             — otherwise it lands on a stale branch and needs a second rebase.
+             Gating it makes a fan-in/chain cascade block on its dependency
+             layers and converge bottom-up (main is always fresh), one rebase
+             per branch. The launching patch's [base_contains_merged_siblings]
+             cache supplies the sibling-containment input. [Respond] operates on
+             a worktree that already exists and is not freshness-sensitive. *)
+          let eligibility patch_id base =
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/orchestrator.ml:338]: Verified safe — this is not a new partiality. \`remove_agent\` purges the outbox for a removed patch, and the pre-existing \`Acked\` branch in \`runnable_messages\` already calls \`agent t msg.patch_id\`. Messages and agents are co-managed, so a Pending message always has its agent. Leaving the internal-invariant accessor as-is, consistent with the surrounding code.

File: lib/patch_controller.ml [hunk 6]
```diff
@@ -112,7 +112,15 @@ let apply_poll_result t patch_id
     if poll_result.merged then (
       let agent = Orchestrator.agent t patch_id in
       if not agent.Patch_agent.merged then log "Merged";
-      Orchestrator.mark_merged t patch_id)
+      let t = Orchestrator.mark_merged t patch_id in
+      (* Record the squash/merge commit SHA so dependents' base-containment gate
+         can ancestry-check it. Only overwrite when present, so a late/absent
+         [mergeCommit.oid] does not clobber a previously recorded SHA. Merged
+         agents that still lack a SHA are kept in the poll loop (see
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/patch_controller.ml:119]: The logic is intentional: \`Some _ as sha\` rebinds the whole option, and skipping the update on \`None\` is deliberate (a late/absent \`mergeCommit.oid\` must not clobber a previously recorded SHA). The bound value is \`string option\`, matching \`set_merge_commit_sha\`'s type, so it type-checks and behaves correctly. Leaving as-is.

File: lib_test/test_generators.ml [hunk 7]
```diff
@@ -471,6 +473,7 @@ let gen_patch_view =
                construct views explicitly rather than relying on this
                generator. *)
             branch_rebased_onto = Some base_branch;
+            base_contains_merged_siblings = true;
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib_test/test_generators.ml:476]: The suggested edit references \`a.Patch_agent.base_contains_merged_siblings\`, but \`gen_patch_view\` builds a \`patch_view\` from a random \`(patch_id, branch)\` pair — there is no agent \`a\` in scope here. The default \`true\` is intentional (non-blocking), matching the existing convention noted in the comment that drift/stale states are exercised with explicitly-constructed views. The stale-sibling (\`false\`) state is covered exhaustively by the new \`test_base_containment.ml\` and the fan-in properties in \`test_reconciler_properties.ml\`.

File: lib_core/base_containment.ml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+         [initial_base] would raise. Report [false] (irrelevant — the patch
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib_core/base_containment.ml:8]: Agreed, no bug — \`_ :: _ :: _\` matches ≥2 elements so exactly-two open deps is caught, and the \`[] | [_]\` arm's \`initial_base\` is total (returns \`main\` for 0 open deps). The "last-but-one-merge edge" comment is only on the \`>1\` arm. Added a totality note in \`187c799\` for the 0-open-deps case.

File: lib_core/base_containment.ml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+         defers on [deps_satisfied] anyway). *)
+      false
+  | [] | [ _ ] ->
+      let base =
+        Graph.initial_base graph patch_id ~has_merged ~branch_of ~main
+      in
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib_core/base_containment.ml:18]: Confirmed correct as you concluded: when one dep is open, the base is that sole open dep's branch and the check is whether each *merged* dep's squash SHA is an ancestor of it — the intended semantics. No change needed.

File: lib_core/base_containment.ml [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+      (* [Graph.initial_base] is total for [<= 1] open dep: [[]] (all deps
+         merged, including a root patch with no deps) resolves to [main]; [[d]]
+         resolves to [d]'s branch. It only raises for [> 1] open deps, which the
+         arm above excludes — so this call never raises. *)
+      let base =
+        Graph.initial_base graph patch_id ~has_merged ~branch_of ~main
+      in
+      if Types.Branch.equal base main then true
+      else
+        Graph.deps graph patch_id |> List.filter ~f:has_merged
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib_core/base_containment.ml:21]: Confirmed and pinned down in this patch. `detect_sibling_stale_bases` only enqueues the sole open dep when `base_rebasable` finds that dep view and `bv.has_pr = true`; if the dep has no PR/worktree yet, `base_rebasable` returns false and no `Rebase` is emitted. I made that contract explicit in `lib_core/reconciler.{ml,mli}` and added `sibling: silent when sole open dep has no PR` to `test/test_reconciler_properties.ml`. The fail-closed ancestry result therefore only defers the fan-in patch until the dep starts/gets a PR or the next successful containment recompute, rather than permanently blocking it.

File: lib/poller_fiber.ml [hunk 11]
```diff
@@ -431,6 +443,52 @@ struct
...
+               deps, so projects with no fan-in-mid-merge pay no git cost here.
+               The oracle is fail-closed: a transient git spawn failure returns
+               [false] rather than aborting the whole reconcile tick (it runs
+               under the runtime lock). Cancellation must still propagate. *)
+            let main_root = lazy (W.resolve_main_root ()) in
+            let ancestor_oracle ancestor ~descendant =
+              try
+                W.is_ancestor ~path:(Lazy.force main_root) ~ancestor ~descendant
+              with
+              | Eio.Cancel.Cancelled _ as exn -> raise exn
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/poller_fiber.ml:478]: Documented this explicitly near the containment-cache fold in `lib/poller_fiber.ml`. The comment now states that cancellation can interrupt after a prefix of agents is updated, that the tick is aborting, and that the next reconcile invocation recomputes every agent using a fresh `main_root` lazy value. The retained cache can therefore only defer Start/Rebase until the next successful pass.

File: lib_core/reconciler.ml [hunk 12]
```diff
@@ -111,6 +120,54 @@ let detect_stale_bases graph views ~has_merged ~branch_of ~main =
...
+    Guards mirror the other detectors (has_pr, ~merged, and skip when [B]
+    already has a [Rebase] queued). The [open_pr_deps = 1] guard keeps
+    [sole_open_dep] total and pins this to the "last-but-one dependency merged"
+    edge, where a single rebase of [B] realizes containment for [P]. *)
+let detect_sibling_stale_bases graph views ~has_merged =
+  let view_by_id =
+    List.fold views
+      ~init:(Map.empty (module Patch_id))
+      ~f:(fun acc v -> Map.set acc ~key:v.id ~data:v)
+  in
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib_core/reconciler.ml:153]: Fixed. `detect_sibling_stale_bases` now builds `view_by_id` with `Map.add` and fails fast on `Duplicate`, including the duplicate `Patch_id` in the error message. This now matches the duplicate-key discipline used by `detect_rebases`.

File: test/test_stale_base_invariants.ml [hunk 13]
```diff
@@ -240,11 +240,19 @@ let sbi_defer_implies_progress m =
...
+          with
           | Start_eligibility.Allow -> true
           | Start_eligibility.Defer
               (Start_eligibility.Base_patch_busy_with_rebase _) ->
               true (* by definition: a Rebase is in flight or queued *)
+          | Start_eligibility.Defer
+              (Start_eligibility.Base_missing_merged_sibling _) ->
+              (* Unreachable: this invariant evaluates the gate with
+                 [~base_contains_merged_siblings:true]. *)
+              true
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_stale_base_invariants.ml]: Fixed. Both unreachable `Base_missing_merged_sibling` branches in `test/test_stale_base_invariants.ml` now return `false`, so an unexpected sibling-deferral reason under `~base_contains_merged_siblings:true` fails the property instead of passing.
